WordPress/gutenberg
# Fix bumped specificity for layout styles in non-iframed editor

Core backport PR - https://github.com/WordPress/wordpress-develop/pull/7145
Core backport trac ticket - https://core.trac.wordpress.org/ticket/61829



## What?
Fixes the issue mentioned in https://github.com/WordPress/gutenberg/issues/53468#issuecomment-2241279024 to restore theme.json spacing rules taking precedence over the implicit spacing rules in a non-iframed editor.

## How?
In `trunk` the code generates a selector like `.is-layout-flow > *` (specificity 0,1,0) which is fine normally. When in a non-iframed editor, the `transformStyles` function prefixes selectors with `.editor-styles-wrapper` to scope them to the editor canvas, so it becomes `.editor-styles-wrapper .is-layout-flow > *` (specificity 0,2,0), a bump in specificity that causes the precedence of styles to change.

In this PR I've changed the selector to `:root :where(.is-layout-flow) > *` (still specificity 0,1,0). `transformStyles` handles 'root' selectors a little differently, it'll instead replace the `:root` part so it becomes `.editor-styles-wrapper where(.is-layout-flow) > *` (keeping the specificity at 0,1,0).

The other layout selector that this affects is the `:first-child` `:last-child` selectors that are responsible for resetting margin at the start and end of a block list. They traditionally have a 0,2,0 specificity so that they can override both the above rule and any rules in the theme.json. Those selectors are also maintained at 0,2,0 with this change, they become something like `:root :where(.is-layout-flow) > :first-child`.

<details><summary>Here's a list of selectors that will be updated (according to our php tests)</summary>

```diff
-.is-layout-flow  > :first-child {
+:root :where(.is-layout-flow) > :first-child {

-.is-layout-flow  > :last-child {
+:root :where(.is-layout-flow) > :last-child {

-.is-layout-flow  > * {
+:root :where(.is-layout-flow) > * {

-.is-layout-constrained  > :first-child {
+:root :where(.is-layout-constrained) > :first-child {

-.is-layout-constrained  > :last-child {
+:root :where(.is-layout-constrained) > :last-child {

-.is-layout-constrained  > * {
+:root :where(.is-layout-constrained) > * {

-.is-layout-flex  {
+:root :where(.is-layout-flex) {

-.is-layout-grid  {
+:root :where(.is-layout-grid) {

-.wp-block-social-links-is-layout-flow  > :first-child {
+:root :where(.wp-block-social-links-is-layout-flow) > :first-child {

-.wp-block-social-links-is-layout-flow  > :last-child {
+:root :where(.wp-block-social-links-is-layout-flow) > :last-child {

-.wp-block-social-links-is-layout-flow  > * {
+:root :where(.wp-block-social-links-is-layout-flow) > * {

-.wp-block-social-links-is-layout-constrained  > :first-child {
+:root :where(.wp-block-social-links-is-layout-constrained) > :first-child {

-.wp-block-social-links-is-layout-constrained  > :last-child {
+:root :where(.wp-block-social-links-is-layout-constrained) > :last-child {

-.wp-block-social-links-is-layout-constrained  > * {
+:root :where(.wp-block-social-links-is-layout-constrained) > * {

-.wp-block-social-links-is-layout-flex  {
+:root :where(.wp-block-social-links-is-layout-flex) {

-.wp-block-social-links-is-layout-grid  {
+:root :where(.wp-block-social-links-is-layout-grid) {

-.wp-block-buttons-is-layout-flow  > :first-child {
+:root :where(.wp-block-buttons-is-layout-flow) > :first-child {

-.wp-block-buttons-is-layout-flow  > :last-child {
+:root :where(.wp-block-buttons-is-layout-flow) > :last-child {

-.wp-block-buttons-is-layout-flow  > * {
+:root :where(.wp-block-buttons-is-layout-flow) > * {

-.wp-block-buttons-is-layout-constrained  > :first-child {
+:root :where(.wp-block-buttons-is-layout-constrained) > :first-child {

-.wp-block-buttons-is-layout-constrained  > :last-child {
+:root :where(.wp-block-buttons-is-layout-constrained) > :last-child {

-.wp-block-buttons-is-layout-constrained  > * {
+:root :where(.wp-block-buttons-is-layout-constrained) > * {

-.wp-block-buttons-is-layout-flex  {
+:root :where(.wp-block-buttons-is-layout-flex) {

-.wp-block-buttons-is-layout-grid  {
+:root :where(.wp-block-buttons-is-layout-grid) {

```

</details> 

## Testing Instructions
1. Add some very obvious margin rules to your theme.json like this for paragraphs (under `styles.blocks`):
```json
"core/paragraph": {
	"spacing": {
		"margin": {
			"top": "4rem",
			"bottom": "4rem"
		}
	}
},
```
2. Create a new post. Ensure you don't have custom fields active.
3. Insert 3 paragraphs with text, check that the margin looks right. The first paragraph should have no top margin, the last no bottom margin, but in all other cases there should be `4rem` top/bottom margin.
4. Activate Custom Fields in the editor preferences to 'un-iframe' the editor
5. Check that the styles are the same without the iframe

Also worth checking out `trunk` and comparing the styles. You should see that in trunk the iframed editor looks correct, but when you activate custom fields the non-iframed editor styles are broken.

## Screenshots or screencast 
#### Before
![Screenshot 2024-07-31 at 4 33 10 PM](https://github.com/user-attachments/assets/1e053fe1-d942-4f71-8a53-ea478b26c0b4)

#### After
![Screenshot 2024-07-31 at 4 33 30 PM](https://github.com/user-attachments/assets/ffb33f68-cf8f-42eb-916c-af4d189d428e)
